File: swift/Sources/Schemas/ClaimView.swift
import Foundation

public struct ClaimView: Codable, Hashable, Sendable {
    public let attachmentId: String
    public let id: String
    public let leaseUntil: Double
    public let ownerEpoch: Int
    public let ownerId: String
    /// Initial launch bound anchored at the first claim request start; idempotent claim retries never restart this clock. Workers cannot renew a ready executor lease.
    public let startupTtlSeconds: Int?
    public let status: String
    /// Additional properties that are not explicitly defined in the schema
    public let additionalProperties: [String: JSONValue]

    public init(
        attachmentId: String,
        id: String,
        leaseUntil: Double,
        ownerEpoch: Int,
        ownerId: String,
        startupTtlSeconds: Int? = nil,
        status: String,
        additionalProperties: [String: JSONValue] = .init()
    ) {
        self.attachmentId = attachmentId
        self.id = id
        self.leaseUntil = leaseUntil
        self.ownerEpoch = ownerEpoch
        self.ownerId = ownerId
        self.startupTtlSeconds = startupTtlSeconds
        self.status = status
        self.additionalProperties = additionalProperties
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        self.attachmentId = try container.decode(String.self, forKey: .attachmentId)
        self.id = try container.decode(String.self, forKey: .id)
        self.leaseUntil = try container.decode(Double.self, forKey: .leaseUntil)
        self.ownerEpoch = try container.decode(Int.self, forKey: .ownerEpoch)
        self.ownerId = try container.decode(String.self, forKey: .ownerId)
        self.startupTtlSeconds = try container.decodeIfPresent(Int.self, forKey: .startupTtlSeconds)
        self.status = try container.decode(String.self, forKey: .status)
        self.additionalProperties = try decoder.decodeAdditionalProperties(using: CodingKeys.self)
    }

    public func encode(to encoder: Encoder) throws -> Void {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try encoder.encodeAdditionalProperties(self.additionalProperties)
        try container.encode(self.attachmentId, forKey: .attachmentId)
        try container.encode(self.id, forKey: .id)
        try container.encode(self.leaseUntil, forKey: .leaseUntil)
        try container.encode(self.ownerEpoch, forKey: .ownerEpoch)
        try container.encode(self.ownerId, forKey: .ownerId)
        try container.encodeIfPresent(self.startupTtlSeconds, forKey: .startupTtlSeconds)
        try container.encode(self.status, forKey: .status)
    }

    /// Keys for encoding/decoding struct properties.
    enum CodingKeys: String, CodingKey, CaseIterable {
        case attachmentId = "attachment_id"
        case id
        case leaseUntil = "lease_until"
        case ownerEpoch = "owner_epoch"
        case ownerId = "owner_id"
        case startupTtlSeconds = "startup_ttl_seconds"
        case status
    }
}

File: rust/src/api/types/claim_view.rs
pub use crate::prelude::*;

#[derive(Debug, Clone, Serialize, Deserialize, Default, PartialEq)]
pub struct ClaimView {
    #[serde(default)]
    pub attachment_id: String,
    #[serde(default)]
    pub id: String,
    #[serde(default)]
    #[serde(with = "crate::core::number_serializers")]
    pub lease_until: f64,
    #[serde(default)]
    pub owner_epoch: i64,
    #[serde(default)]
    pub owner_id: String,
    /// Initial launch bound anchored at the first claim request start; idempotent claim retries never restart this clock. Workers cannot renew a ready executor lease.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub startup_ttl_seconds: Option<i64>,
    #[serde(default)]
    pub status: String,
}

impl ClaimView {
    pub fn builder() -> ClaimViewBuilder {
        <ClaimViewBuilder as Default>::default()
    }
}

#[derive(Clone, PartialEq, Default, Debug)]
#[non_exhaustive]
pub struct ClaimViewBuilder {
    attachment_id: Option<String>,
    id: Option<String>,
    lease_until: Option<f64>,
    owner_epoch: Option<i64>,
    owner_id: Option<String>,
    startup_ttl_seconds: Option<i64>,
    status: Option<String>,
}

impl ClaimViewBuilder {
    pub fn attachment_id(mut self, value: impl Into<String>) -> Self {
        self.attachment_id = Some(value.into());
        self
    }

    pub fn id(mut self, value: impl Into<String>) -> Self {
        self.id = Some(value.into());
        self
    }

    pub fn lease_until(mut self, value: f64) -> Self {
        self.lease_until = Some(value);
        self
    }

    pub fn owner_epoch(mut self, value: i64) -> Self {
        self.owner_epoch = Some(value);
        self
    }

    pub fn owner_id(mut self, value: impl Into<String>) -> Self {
        self.owner_id = Some(value.into());
        self
    }

    pub fn startup_ttl_seconds(mut self, value: i64) -> Self {
        self.startup_ttl_seconds = Some(value);
        self
    }

    pub fn status(mut self, value: impl Into<String>) -> Self {
        self.status = Some(value.into());
        self
    }

    /// Consumes the builder and constructs a [`ClaimView`].
    /// This method will fail if any of the following fields are not set:
    /// - [`attachment_id`](ClaimViewBuilder::attachment_id)
    /// - [`id`](ClaimViewBuilder::id)
    /// - [`lease_until`](ClaimViewBuilder::lease_until)
    /// - [`owner_epoch`](ClaimViewBuilder::owner_epoch)
    /// - [`owner_id`](ClaimViewBuilder::owner_id)
    /// - [`status`](ClaimViewBuilder::status)
    pub fn build(self) -> Result<ClaimView, BuildError> {
        Ok(ClaimView {
            attachment_id: self
                .attachment_id
                .ok_or_else(|| BuildError::missing_field("attachment_id"))?,
            id: self.id.ok_or_else(|| BuildError::missing_field("id"))?,
            lease_until: self
                .lease_until
                .ok_or_else(|| BuildError::missing_field("lease_until"))?,
            owner_epoch: self
                .owner_epoch
                .ok_or_else(|| BuildError::missing_field("owner_epoch"))?,
            owner_id: self
                .owner_id
                .ok_or_else(|| BuildError::missing_field("owner_id"))?,
            startup_ttl_seconds: self.startup_ttl_seconds,
            status: self
                .status
                .ok_or_else(|| BuildError::missing_field("status"))?,
        })
    }
}
